src/MySqlConnector/MySqlClient/ConnectionPool.cs

```diff
@@ -14,6 +14,13 @@ internal sealed class ConnectionPool
 		{
 			cancellationToken.ThrowIfCancellationRequested();
 
+			// if all sessions are used, see if any have been leaked and can be recovered
+			if (m_sessionSemaphore.CurrentCount == 0)
+			{
+				GC.Collect();
+				RecoverLeakedSessions();
```

Comment:
This could happen potentially thousands of times in a second, and calls locked code. Maybe adapt it to run at most once per second?